lib/ecto/repo/schema.ex

```diff
@@ -91,28 +93,33 @@ defmodule Ecto.Repo.Schema do
     end
   end
 
-  defp init_mapper(nil, _dumper, _adapter) do
+  defp init_mapper(nil, _dumper, _adapter, _placeholder_map) do
     fn {field, _} = tuple, {header, has_query?} ->
       {tuple, {Map.put(header, field, true), has_query?}}
     end
   end
 
-  defp init_mapper(schema, dumper, adapter) do
+  defp init_mapper(schema, dumper, adapter, placeholder_map) do
     fn {field, value}, {header, has_query?} ->
-        case dumper do
-          %{^field => {source, type}} ->
-            case value do
-              %Ecto.Query{} = query ->
-                {{source, query}, {Map.put(header, source, true), true}}
-
-              value ->
-                value = dump_field!(:insert_all, schema, field, type, value, adapter)
-                {{source, value}, {Map.put(header, source, true), has_query?}}
-            end
-          %{} ->
-            raise ArgumentError, "unknown field `#{inspect(field)}` in schema #{inspect(schema)} given to " <>
-                                 "insert_all. Note virtual fields and associations are not supported"
-        end
+      case dumper do
+        %{^field => {source, type}} ->
+          case value do
+            %Ecto.Query{} = query ->
+              {{source, query}, {Map.put(header, source, true), true}}
+
+            {:placeholder, placeholder_key} ->
+              replaced_value = Map.fetch!(placeholder_map, placeholder_key)
+              value = dump_field!(:insert_all, schema, field, type, replaced_value, adapter)
+              {{source, value}, {Map.put(header, source, true), has_query?}}
```

Comment:
I believe you want to return `{:source, {:placeholder, placeholder_key}}` and instead update the dumped value in the placeholder map itself.

Reply:
To be clear, I think it will be simpler if we update the to the placeholder map in here and pass the updated placeholder map to the `adapter.insert_all` call here: https://github.com/elixir-ecto/ecto/pull/3515/files#diff-91ed3e8c550063fcf5e2df36020d92405a303c5f5745c0bf643c7dc4c538befbR56

Then there is less reconstruction to do on the `ecto_sql` side of things.